Update leave_policy/parallel phase-related RFCs (#2130)
oneapi-src/oneTBB: rfcs/experimental/parallel_phase_for_task_arena/README.md

Reviewer: This is more a process question. We are aware of improvements, but there's nothing documented in the RFC. Should there be?

Author: Perhaps we could add at least some plan on measuring performance. I was thinking about updating the seismic example, since it showcases sequential composability scenario.

Author: Added "Performance testing" section with the suggestion to use seismic to measure performance.

Reviewer 2: I'd change wording to "implementation-specific", because in certain contexts (e.g., the C++ standard) "implementation-defined" means "chosen by implementors but clearly documented and guaranteed".

Author: Done.

<img src="completely_disable_new_behavior.png" width=800>

> The question is whether we should allow the arena to dynamically transition from one state to

Reviewer: We've decided to not support dynamic transition at this point. So I think you should be more clear that this is a decision and no longer an open question. We can revisit if additional use cases arise (which you can state). But I'd suggest being more clear that for production, this is decided.

Essentially make the language stronger, removing things like "at this point" and maybe start with "One question was whether".

Author: Updated.

another. Due to the lack of use cases that would justify this, the decision is to keep the state
static for the lifetime of the arena. The current implementation is flexible enough to support such

> extension (see [technical details](#technical-details)).

Reviewer 2: It's not fully clear to me which extension do you mean here.

Author: By "extension" I meant extending the functionality around `leave_policy` setting to allow changing it dynamically.

Reviewer 2: Maybe you can adjust the wording to make that clear.

Author: Done.

                    priority a_priority = priority::normal,
                    leave_policy a_leave_policy = leave_policy::automatic);

    class parallel_phase {
        enum class end_flags : /* unspecified type */ {
            with_fast_leave = /* unspecified */
        };
        parallel_phase(attach, end_flags e_flags = {});
        parallel_phase(task_arena& ta, end_flags e_flags = {});
    };
    
    void start_parallel_phase();
    void end_parallel_phase(parallel_phase::end_flags e_flags = {});
};

namespace this_task_arena {
    void start_parallel_phase();
    void end_parallel_phase(task_arena::parallel_phase::end_flags e_flags = {});
}
```
The _parallel phase_ continues until each previous `start_parallel_phase` call
to the same arena has a matching `end_parallel_phase` call.
Let's also introduce RAII object `parallel_phase` that will help to manage the contract.
Rather than introducing a separate RAII type in the `this_task_arena` namespace,
`task_arena::parallel_phase` can be constructed with the `tbb::attach` tag to map the phase to

> the implicit arena associated with the calling thread.

Author: This part relates to the open question about RAII parallel_phase object for "this_task_arena" ([see discussion](https://github.com/uxlfoundation/oneTBB/pull/1522#discussion_r1871955641)).

Author: This question should be resolved by choosing the way to create RAII `parallel_phase` object (attach parameter vs using factory function).

Note that `end_parallel_phase` and the `parallel_phase` constructor accept an `end_flags`
argument rather than a boolean flag even though `end_flags` has a single flag. Since _parallel phase_
is a high-level hint to the scheduler, it makes sense that other scheduling hints could be tied to
it as well, so the API is designed to be configurable. A symmetric `start_flags` parameter
could be added to `start_parallel_phase` in the future if needed.

If the end of the parallel phase is not indicated by the user, it will be done automatically when
the last public reference is removed from the arena (i.e., task_arena has been destroyed or,
for an implicitly created arena, the thread that owns it has completed).
This ensures correctness is preserved (threads will not be retained forever).

> Introduction of the explicit functions `start_parallel_phase` and `end_parallel_phase` opens

Author: This part with the examples tries to figure out the necessity of explicit "start/end_parallel_phase" functions vs. limiting the API only to RAII class ([see discussion](https://github.com/uxlfoundation/oneTBB/pull/1522#discussion_r1871505223)).

a possibility of misuse: either forgetting to pair phase starts with ends, which is not a
correctness issue per se as described above, or doing more phase ends than starts
(e.g. by using RAII class interchangeably with explicit calls).
That raises a question of whether the parallel phase API should be limited only to the RAII style.
To answer this question, we need to consider the use cases where explicit calls might be more preferable.

One such use case is asynchronous handoff, where the start and the end of a parallel phase happen
in different scopes, potentially executed on different threads, so there is no single scope
to attach an RAII guard to:
```cpp
void handle_request(Request req) {
    tbb::this_task_arena::start_parallel_phase();
    //
    // Some composition of parallel and serial computations
    //
    tbb::this_task_arena::enqueue([req]() {
        process(req);
        tbb::this_task_arena::end_parallel_phase(tbb::task_arena::parallel_phase::end_flags::with_fast_leave);
    });
}
```

The same use case can be expressed using the `parallel_phase` object, assuming move construction is supported:

```cpp
void handle_request(Request req) {

>     tbb::task_arena::parallel_phase phase{tbb::attach{}};

Reviewer 3: ```suggestion
void handle_request(Request req) {
    tbb::task_arena::parallel_phase phase{tbb::attach{},
                                          tbb::task_arena::parallel_phase::end_flags::with_fast_leave};
```

    });
}
```
Another use case is a long-lived service where the parallel phase is meant to span multiple,
unrelated requests and should only end after a period of inactivity, e.g. driven by an idle timer
rather than the destruction of a scope:
```cpp
class Service {
    tbb::task_arena ta;
    bool phase_started = false;

    void on_request() {
        if (!phase_started) {
            ta.start_parallel_phase();
            phase_started = true;
        }
        ta.execute([]() { /* work */ });
        reset_idle_timer();
    }

    void on_idle_timeout() {
        ta.end_parallel_phase();
        phase_started = false;
    }
};
```
This use case can also be expressed using the `parallel_phase` object:
```cpp
class Service {
    tbb::task_arena ta;
    std::optional<tbb::task_arena::parallel_phase> phase;

    void on_request() {
        if (!phase) {
            phase.emplace(ta);
        }
        ta.execute([]() { /* work */ });
        reset_idle_timer();
    }

    void on_idle_timeout() {
        phase.reset();
    }
};
```
If the friction of dealing with the RAII object is considered too high, then it makes sense to provide the explicit functions as well.

> The start of _parallel phase_ can also be used as a warm-up hint for the workers to enter

Author: And this paragraph is about the expectations from parallel_phase API when a calling thread doesn't yet have an associated arena with it.

                    priority a_priority = priority::normal,
                    leave_policy a_leave_policy = leave_policy::automatic);

    class parallel_phase {
        enum class end_flags : /* unspecified type */ {
            with_fast_leave = /* unspecified */
        };
        parallel_phase(attach, end_flags e_flags = {});
        parallel_phase(task_arena& ta, end_flags e_flags = {});
    };
    
    void start_parallel_phase();
    void end_parallel_phase(parallel_phase::end_flags e_flags = {});
};

namespace this_task_arena {
    void start_parallel_phase();

>     void end_parallel_phase(task_arena::parallel_phase::end_flags e_flags = {});

Reviewer 4: This seems to change the default from `false` to `true` and makes an ending of a parallel phase using anything besides `with_fast_leave` value impossible. Is it what is intended?

Author: Not necessarily. The value of `with_fast_leave` is unspecified, so internally it can be set any non-zero value.

Reviewer 4: Can we use `task_arena::leave_policy` type instead?

Consider `void end_parallel_phase(task_arena::leave_policy end_phase_policy)`.

I think introducing separate `end_flags` type for essentially the same entity is not needed.

Author: [Again, parallel_phase is not about selecting `leave_policy` for task arena.](https://github.com/uxlfoundation/oneTBB/pull/1522#discussion_r1861130746)

                    priority a_priority = priority::normal,
                    leave_policy a_leave_policy = leave_policy::automatic);

    class parallel_phase {
        // Available only when each type in Flags is a parallel phase flag
        class flags {
            template <typename... Flags>
            flags(Flags... f);
        };
        class end_with_fast_leave;
        parallel_phase(attach, flags f = {});
        parallel_phase(task_arena& ta, flags f = {});
        void end();
    };
    
    void start_parallel_phase(parallel_phase::flags f = {});
    void end_parallel_phase(parallel_phase::flags f = {});
};

namespace this_task_arena {
    void start_parallel_phase(task_arena::parallel_phase::flags f = {});
    void end_parallel_phase(task_arena::parallel_phase::flags f = {});
}
```
The _parallel phase_ continues until each previous `start_parallel_phase` call
to the same arena has a matching `end_parallel_phase` call.

> Let's also introduce RAII object `parallel_phase` that will help to manage the contract.

Reviewer: ```suggestion
Let's also introduce an RAII object `parallel_phase` that will help to manage the contract.
```

Rather than introducing a separate RAII type in the `this_task_arena` namespace,
`task_arena::parallel_phase` can be constructed with the `tbb::attach` tag to map the phase to
the implicit arena associated with the calling thread. The alternative would be, instead of
passing `tbb::attach`, to have a factory function that returns a `pararlel_phase` object:
```cpp
class task_arena {
    parallel_phase create_parallel_phase(parallel_phase::flags f = {});
};

namespace this_task_arena {
    task_arena::parallel_phase create_parallel_phase(task_arena::parallel_phase::flags f = {});
}
```

> The approach is more consistent with the rest of task arena API.

Reviewer: Which one is "this" approach? You've just described two alternatives.

Author: Clarified, I meant the factory function approach.

> The approach is more consistent with the rest of task arena API.

Author: ```suggestion
The factory function approach seems more consistent with the rest of task arena API.
```

        ta.end_parallel_phase();
        phase_started = false;
    }
};
```
This use case can also be expressed using the `parallel_phase` object:
```cpp
class Service {
    tbb::task_arena ta;
    std::unique_ptr<tbb::task_arena::parallel_phase> phase;

    void on_request() {
        if (!phase) {
            phase = std::make_unique<tbb::task_arena::parallel_phase>(ta);
            // or
            // phase = std::make_unique<tbb::task_arena::parallel_phase>(ta.create_parallel_phase());
        }
        ta.execute([]() { /* work */ });
        reset_idle_timer();
    }

    void on_idle_timeout() {
        phase.reset();
    }
};
```
If the friction of dealing with the RAII object is considered too high, then it makes sense to provide the explicit functions as well.

The start of _parallel phase_ can also be used as a warm-up hint for the workers to enter
the arena in advance, therefore the arena (including the implicit arena bound to the external thread)
must be initialized during the first call to `start_parallel_phase`. It means that if a calling thread has
no associated arena yet, the invocation of `this_task_arena::start_parallel_phase` will initialize

> an arena and bind it to the calling thread.

Reviewer: This equally would apply to `this_task_arena::create_parallel_phase` not just `start_parallel_phase`, right? So this is not a deciding factor in choosing RAII object over explicit start/stop functions.

Author: Yes, it is actually a separate question elaborated here. Maybe it is better to separate them by subheaders.

Author: Separated by subheaders.

                    priority a_priority = priority::normal,
                    leave_policy a_leave_policy = leave_policy::automatic);

    class parallel_phase {
        // Available only when each type in Flags is a parallel phase flag
        class flags {
            template <typename... Flags>
            flags(Flags... f);
        };
        class end_with_fast_leave;

>         parallel_phase(attach, flags f = {});

Reviewer 2: Do we really need the `attach` flag here? An alternative is to assume by default that the parallel phase acts in the current arena of a thread; would it have some downsides compared to the explicit flag?

Author: I prefer having the tag because it is a bit more descriptive and in the context of task_arena `attach` tag already means that the instance is associated with the arena of a calling thread.

Reviewer 2: OK.

Some flags are only applicable to the start of a parallel phase, and others only to its end, which
can be reflected in their names (e.g. `end_with_fast_leave`). The `parallel_phase` object accepts flags
for both boundaries at once and applies each of them at the corresponding point, while the explicit
functions only take into account the flags applicable to them. A flag passed to an operation it
does not apply to is ignored. 

#### Starting and ending a parallel phase

The start of _parallel phase_ can also be used as a warm-up hint for the workers to enter
the arena in advance, therefore the arena (including the implicit arena bound to the external thread)
must be initialized during the first call to `start_parallel_phase`. It means that if a calling thread has
no associated arena yet, the invocation of `this_task_arena::start_parallel_phase` will initialize
an arena and bind it to the calling thread.

If the end of the parallel phase is not indicated by the user, it will be done automatically when
the last public reference is removed from the arena (i.e., task_arena has been destroyed or,
for an implicitly created arena, the thread that owns it has completed).

> This ensures correctness is preserved (threads will not be retained forever).

Reviewer 2: What if the `end_parallel_phase` or `parallel_phase` destructor is called for an arena that is already destroyed?

Author: It seems like you won't be able to do so without creating a dangling reference to `task_arena` which seems more like "you already lost" situation. `enter/exit_parallel_phase` entry points accept `task_arena_base*` and if it is `nullptr`, arena bound to a calling thread is used.
If user explicitly calls `terminate` before ending the phase, while keeping the task arena object, we can add an assertion that during the phase end, the arena is still initialized.
Interesting thing is that, since we don't propagate `r1::arena` back, you can't start a parallel_phase for `this_task_arena` in one arena and end it in another.

Reviewer 2: If it is a requirement that the lifetime of the arena object cannot end while a parallel phase is active, it should be stated as such. And then, "the end of the parallel phase is not indicated by the user" is not even a concern but simply undefined behavior.

Author: Updated.

## Open Questions in Design

Some open questions that remain:

> * Are there additional use cases that should be considered that we missed in our analysis?

Reviewer: I'd remove this. As we move towards productization, we currently are not aware of any. If they arise, they become a new feature request later.

Author: Done.
